coded_tools/multiagent_decomposer/decomposition_solver.py

```diff
@@ -0,0 +1,144 @@
+        scope of your CodedTool and can be done asynchronously, especially within
+        the context of your CodedTool running within a server.
+
+        If you find your CodedTools can't help but synchronously block,
+        strongly consider looking into using the asyncio.to_thread() function
+        to not block the EventLoop for other requests.
+        See: https://docs.python.org/3/library/asyncio-task.html#asyncio.to_thread
+        Example:
+            async def async_invoke(self, args: Dict[str, Any], sly_data: Dict[str, Any]) -> Any:
+                return await asyncio.to_thread(self.invoke, args, sly_data)
+
+        :param args: An argument dictionary whose keys are the parameters
+                to the coded tool and whose values are the values passed for them
+                by the calling agent.  This dictionary is to be treated as read-only.
+        :param sly_data: A dictionary whose keys are defined by the agent hierarchy,
+                but whose values are meant to be kept out of the chat stream.
+
+                This dictionary is largely to be treated as read-only.
+                It is possible to add key/value pairs to this dict that do not
+                yet exist as a bulletin board, as long as the responsibility
+                for which coded_tool publishes new entries is well understood
+                by the agent chain implementation and the coded_tool implementation
+                adding the data is not invoke()-ed more than once.
+        :return: A return value that goes into the chat stream.
+        """
+
+        # Create the solver and use some of the arguments to configure it
+        solver = NeuroSanSolver(winning_vote_count=args.get("winning_vote_count", 2),
+                                candidate_count=args.get("candidate_count"),
+                                number_of_votes=args.get("number_of_votes"),
+                                solution_candidate_count=args.get("solution_candidate_count"))
+
+        tools: Dict[str, str] = {}
+        tools = args.get("tools", tools)
+
+        # Set up the AgentCallers to use this CodedTool as a basis for calling the agents.
+        parsing = SolverParsing()
+        composition_discriminator_caller = CodedToolAgentCaller(self, parsing,
+                                                                name=tools.get("composition_discriminator"))
+        decomposer_caller = CodedToolAgentCaller(self, parsing=None, name=tools.get("decomposer"))
+        problem_solver_caller = CodedToolAgentCaller(self, parsing=None, name=tools.get("problem_solver"))
+        solution_discriminator_caller = CodedToolAgentCaller(self, parsing, name=tools.get("solution_discriminator"))
+        solver.set_callers(composition_discriminator_caller,
+                           decomposer_caller,
+                           problem_solver_caller,
+                           solution_discriminator_caller)
```

Comment:
Set up a bunch of CodedToolAgentCallers so that the NeuroSanSolver can use this object's use_tool() API from neuro-san to call the agents.